lib/ClusterShell/Worker/Rsh.py

```diff
@@ -59,8 +61,31 @@ def _build_cmd(self):
         cmd_l.append("%s" % self.key)  # key is the node
         cmd_l.append("%s" % self.command)
 
+        # Save the return code from the command if required
+        if maxrc:
+            cmd_l.append("; echo XXRETCODE: $?")
+
         return (cmd_l, None)
 
+    def _parse_line(self, line, sname):
+
+        # Read the return code from the command if required
+        task = self.worker.task
+        maxrc = task.info("maxrc", False)
+        if maxrc:
+            match = re.search("^XXRETCODE: (\d+)$", line.decode("utf-8"))
+            if match:
+                self.worker.current_rc = int(match.group(1))
+            else:
+                self.worker._on_node_msgline(self.key, line, sname)
+        else:
+            self.worker._on_node_msgline(self.key, line, sname)
+
+    def _handle_read(self, sname):
```

Comment:
I think you should override `_on_nodeset_msgline()` instead, because this is called from `_handle_read()` and `_flush_read()` and you are missing `_flush_read()` right now.

You just need to search for the pattern, then, call `_on_nodeset_msgline()`. Store the rc in a client local variable. Then override `_on_nodeset_close()` and use that value.
This function could be called with either the actual rc from rsh (0 or 1) or 128 - SIGNUM is the rsh process was killed by a signal. You should probably check for 0 or 1 and use the saved rc only in this case. Mmh... actually, the local rc would be None in this case, so that should not be difficult anyway...

Reply:
Thanks @degremont.  I believe I have made your requested changes, though I was a little confused when you said I should override `_on_nodeset_msgline()` and pithing there I should call `_on_nodeset_msgline ()`.  I am not 100% sure, but I believe you meant that I should call `_on_node_msgline()` from within the overridden function.  Please let me know if I misunderstood.  Thanks!